plutus-tx/src/PlutusTx/Ratio.hs

```diff
@@ -4,80 +4,349 @@
+
+instance P.Ord Rational where
+  {-# INLINEABLE compare #-}
+  compare (Rational n d) (Rational n' d') = P.compare (n P.* d') (n' P.* d)
+  {-# INLINEABLE (<=) #-}
+  Rational n d <= Rational n' d' = (n P.* d') P.<= (n' P.* d)
+  {-# INLINEABLE (>=) #-}
+  Rational n d >= Rational n' d' = (n P.* d') P.>= (n' P.* d)
+  {-# INLINEABLE (<) #-}
+  Rational n d < Rational n' d' = (n P.* d') P.< (n' P.* d)
+  {-# INLINEABLE (>) #-}
+  Rational n d > Rational n' d' = (n P.* d') P.> (n' P.* d)
+
+instance Ord Rational where
+  compare (Rational n d) (Rational n' d') = compare (n * d') (n' * d)
+  Rational n d <= Rational n' d' = (n * d') <= (n' * d)
+  Rational n d >= Rational n' d' = (n * d') >= (n' * d)
+  Rational n d < Rational n' d' = (n * d') < (n' * d)
+  Rational n d > Rational n' d' = (n * d') > (n' * d)
+
+instance P.AdditiveSemigroup Rational where
+  {-# INLINEABLE (+) #-}
+  Rational n d + Rational n' d' =
+    let newNum = (n P.* d') P.+ (n' P.* d)
+        newDen = d P.* d'
+        gcd' = euclid newNum newDen
+     in Rational (newNum `Builtins.quotientInteger` gcd')
+                 (newDen `Builtins.quotientInteger` gcd')
+
+instance P.AdditiveMonoid Rational where
+  {-# INLINEABLE zero #-}
+  zero = Rational P.zero P.one
+
+instance P.AdditiveGroup Rational where
+  {-# INLINEABLE (-) #-}
+  Rational n d - Rational n' d' =
+    let newNum = (n P.* d') P.- (n' P.* d)
+        newDen = d P.* d'
+        gcd' = euclid newNum newDen
+     in Rational (newNum `Builtins.quotientInteger` gcd')
+                 (newDen `Builtins.quotientInteger` gcd')
+
+instance P.MultiplicativeSemigroup Rational where
+  {-# INLINEABLE (*) #-}
+  Rational n d * Rational n' d' =
+    let newNum = n P.* n'
+        newDen = d P.* d'
+        gcd' = euclid newNum newDen
+     in Rational (newNum `Builtins.quotientInteger` gcd')
+                 (newDen `Builtins.quotientInteger` gcd')
+
+instance P.MultiplicativeMonoid Rational where
+  {-# INLINEABLE one #-}
+  one = Rational P.one P.one
+
+instance P.Module Integer Rational where
+  {-# INLINEABLE scale #-}
+  scale i (Rational n d) = let newNum = i P.* n
+                               gcd' = euclid newNum d in
+    Rational (newNum `Builtins.quotientInteger` gcd')
+             (d `Builtins.quotientInteger` gcd')
+
+instance P.ToData Rational where
+  {-# INLINEABLE toBuiltinData #-}
+  toBuiltinData (Rational n d) = P.toBuiltinData (n, d)
+
+-- These instances ensure that the following invariants don't break:
+--
+-- 1. The denominator is greater than 0; and
+-- 2. The numerator and denominator are coprime.
+--
+-- For invariant 1, fromBuiltinData yields Nothing on violation, while
+-- unsafeFromData calls error. Invariant 2 is kept maintained by use of
+-- unsafeRatio.
+
+instance P.FromData Rational where
+  {-# INLINEABLE fromBuiltinData #-}
+  fromBuiltinData dat = case P.fromBuiltinData dat of
+    P.Nothing -> P.Nothing
+    P.Just (n, d) -> if d P.== P.zero
```

Comment:
No, because that would `error`. I have to trap the zero-denominator case to give a `Nothing` instead.

Reply:
Really?  It says
```
P.Just (n, d) -> if d P.== P.zero
                        then Builtins.error ()
                        else P.Just (unsafeRatio n d)
```
I thought that if it said
```
P.Just (n, d) ->  P.Just (unsafeRatio n d)
```
then that would do the same thing because `unsafeRatio` will call `error` in the zero case.  However, (a) I'm not sure how that would affect the compiled Plutus Core, and (b) it's probably better to be explicit about what's going on.  I don't think it does any harm doing it this way, so just ignore me.

Reply:
You're not returning `Nothing` on line 150, you're calling error! So I think Kenneth is right that the current code could be replaced by `unsafeRatio`, but perhaps you *meant* to return `Nothing` there?